### PR TITLE
Update docs/arch.md documentation to mention the new gatherers

### PR DESCRIPTION
**Pull request by tremes**
## Categories

- [ ] Bugfix
- [ ] Enhancement
- [ ] Backporting
- [X] Others (CI, Infrastructure, Documentation)

## Sample Archive

No new data

## Documentation

- `docs/arch.md` updated

## Unit Tests

No new test

## Privacy

Yes. There are no sensitive data in the newly collected information.

## Breaking Changes

No

## References

https://issues.redhat.com/browse/CCXDEV-6176
https://bugzilla.redhat.com/show_bug.cgi?id=???
https://access.redhat.com/solutions/???

### CODE REVIEW

File: docs/arch.md [hunk 1]
```diff
@@ -285,30 +285,28 @@ Such a client is used in [GatherMachineSet](pkg/gather/clusterconfig/clusterconf
...
+
+The data from this gatherer is stored under `/config` directory in the archive.
+
+### Workloads gatherer
+
+Defined in [workloads_gatherer.go](https://github.com/openshift/insights-operator/blob/master/pkg/gatherers/workloads/workloads_gatherer.go). This gatherer runs only each 12 hours and the interval is not configurable. This is because it means significantly more data in the archive, and it is also assumed that this data will not change as often. There's only one gathering function in this gatherer and it gathers workloads fingerprint data (SHA of the images, fingerprints of namespaces as number of pods in namespace, fingerprints of containers as first command and first argument).
+
+The data from this gatherer is stored in `/config/workload_info.json` file in the archive, but please note that not every archive contains this data.
```
Comment (xJustin) [inline docs/arch.md:303]:
```suggestion
The data from this gatherer is stored in the `/config/workload_info.json` file in the archive, but please note that not every archive contains this data. 
```

File: docs/arch.md [hunk 2]
```diff
@@ -285,30 +285,28 @@ Such a client is used in [GatherMachineSet](pkg/gather/clusterconfig/clusterconf
...
+
+The data from this gatherer is stored in `/config/workload_info.json` file in the archive, but please note that not every archive contains this data. 
+
+### Conditional gatherer
+
+Defined in [conditional_gatherer.go](https://github.com/openshift/insights-operator/blob/master/pkg/gatherers/conditional/conditional_gatherer.go). This gatherer is run regularly (2h by default), but it gathers some data only when a corresponding condition is met. The conditions and corresponding gathering functions are defined as [a slice of gathering rules](https://github.com/openshift/insights-operator/blob/master/pkg/gatherers/conditional/conditional_gatherer.go#L62). A typical example of a condition is when an alert is firing. This also means that this gatherer relies on availability of Prometheus metrics and alerts. 
+
+The data from this gatherer is stored under `/conditional` directory in the archive.
```
Comment (xJustin) [inline docs/arch.md:309]:
```suggestion
The data from this gatherer is stored under the `/conditional` directory in the archive. 
```

**Review by xJustin — COMMENTED** [review]
@tremes just some minor grammar suggestions. If you feel like it changes the meaning too much, you can ignore.

File: docs/arch.md [hunk 3]
```diff
@@ -285,30 +285,28 @@ Such a client is used in [GatherMachineSet](pkg/gather/clusterconfig/clusterconf
...
-- `important` meaning if that gather-function has an error we will notify `periodic.go` about it, which will handle it accordingly.
-- `failable` meaning if that gather-function has an error we will just log it and add it to our metadata.
-This is necessary as we are expanding into gathering data about resources that are not guaranteed to be present on the cluster. By default if a resource is not present we shouldn't see an error, but it's better to be safe.
+Every gatherer must implement the [Interface](https://github.com/openshift/insights-operator/blob/master/pkg/gatherers/interface.go#L11) and there are three types of gatherers (see below). These are initialized by calling [gather.CreateAllGatherers](https://github.com/openshift/insights-operator/blob/master/pkg/controller/operator.go#L124) in `operator.go`. The actual gathering is triggered in [periodic.go](https://github.com/openshift/insights-operator/blob/master/pkg/controller/periodic/periodic.go#L93), but not every gatherer is triggered every time (see for example the [CustomPeriodGatherer type](https://github.com/openshift/insights-operator/blob/master/pkg/gatherers/interface.go#L21)).
```
[the remark this answers is not in the document]
Comment (tremes) [reply docs/arch.md]: Thanks a lot. I used most of the suggestions here.

File: docs/arch.md [hunk 4]
```diff
@@ -285,30 +285,28 @@ Such a client is used in [GatherMachineSet](pkg/gather/clusterconfig/clusterconf
...
+One of the attributes of the `GatheringClosure` type is the actual function with following return values `([]record.Record, []error)`. The slice of records is the result of gathering function. The actual data is in `Item` attribute of the `Record`. This `Item` is of type [Marshalable](https://github.com/openshift/insights-operator/blob/master/pkg/record/record.go#L32) and there are two JSON marshallers used to serialize the data - [JSONMarshaller](https://github.com/openshift/insights-operator/blob/master/pkg/record/record.go#L37) and [ResourceMarshaller](https://github.com/openshift/insights-operator/blob/master/pkg/record/resource_marshaller.go#L19) which allows you to save few bytes by omitting the `managedFields` during the serialization.
```
[the remark this answers is not in the document]
Comment (tremes) [reply docs/arch.md]: Thanks a lot. I used most of the suggestions here.

File: docs/arch.md [hunk 5]
```diff
@@ -285,30 +285,28 @@ Such a client is used in [GatherMachineSet](pkg/gather/clusterconfig/clusterconf
...
+
+### Clusterconfig gatherer
+
+Defined in [clusterconfig_gatherer.go](https://github.com/openshift/insights-operator/blob/master/pkg/gatherers/clusterconfig/clusterconfig_gatherer.go). This gatherer is run regularly (2h by default) and gathers various data related to cluster config (see [gathered-data doc](https://github.com/openshift/insights-operator/blob/master/docs/gathered-data.md) for more details).
+
+The data from this gatherer is stored under `/config` directory in the archive.
+
+### Workloads gatherer
+
+Defined in [workloads_gatherer.go](https://github.com/openshift/insights-operator/blob/master/pkg/gatherers/workloads/workloads_gatherer.go). This gatherer runs only each 12 hours and the interval is not configurable. This is because it means significantly more data in the archive, and it is also assumed that this data will not change as often. There's only one gathering function in this gatherer and it gathers workloads fingerprint data (SHA of the images, fingerprints of namespaces as number of pods in namespace, fingerprints of containers as first command and first argument).
```
[the remark this answers is not in the document]
Comment (tremes) [reply docs/arch.md]: Thanks a lot. I think I used all the suggestions here. :)

File: docs/arch.md [hunk 6]
```diff
@@ -285,30 +285,28 @@ Such a client is used in [GatherMachineSet](pkg/gather/clusterconfig/clusterconf
...
+Every gatherer must implement the [Interface](https://github.com/openshift/insights-operator/blob/master/pkg/gatherers/interface.go#L11) and there are three types of gatherers (see below). These are initialized by calling [gather.CreateAllGatherers](https://github.com/openshift/insights-operator/blob/master/pkg/controller/operator.go#L124) in `operator.go`. The actual gathering is triggered in [periodic.go](https://github.com/openshift/insights-operator/blob/master/pkg/controller/periodic/periodic.go#L93), but not every gatherer is triggered every time (see for example the [CustomPeriodGatherer type](https://github.com/openshift/insights-operator/blob/master/pkg/gatherers/interface.go#L21)).
+
+Every gatherer can have more gathering functions. Gathering functions are defined as a map, when key is the name of the function and value is [GatheringClosure type](https://github.com/openshift/insights-operator/blob/master/pkg/gatherers/interface.go#L34) and they are executed concurrently in [HandleTasksConcurrently function](https://github.com/openshift/insights-operator/blob/master/pkg/gather/tasks_processing.go#L33) in `task_processing.go`.
```
[the remark this answers is not in the document]
Comment (tremes) [reply docs/arch.md]: Thanks a lot. I used most of the suggestions here.

**Review by natiiix — APPROVED** [review]
Seems alright to me. Definitely better than the outdated information we had. Thanks!
/lgtm

File: docs/arch.md [hunk 7]
```diff
@@ -285,30 +285,28 @@ Such a client is used in [GatherMachineSet](pkg/gather/clusterconfig/clusterconf
...
-- `failable` meaning if that gather-function has an error we will just log it and add it to our metadata.
-This is necessary as we are expanding into gathering data about resources that are not guaranteed to be present on the cluster. By default if a resource is not present we shouldn't see an error, but it's better to be safe.
+Insights operator defines three types of gatherers (see below). Each of them must implement the [Interface](https://github.com/openshift/insights-operator/blob/master/pkg/gatherers/interface.go#L11) and there are three types of gatherers (see below). These are initialized by calling [gather.CreateAllGatherers](https://github.com/openshift/insights-operator/blob/master/pkg/controller/operator.go#L124) in `operator.go`. The actual gathering is triggered in [periodic.go](https://github.com/openshift/insights-operator/blob/master/pkg/controller/periodic/periodic.go#L93), but not every gatherer is triggered every time ( for example, see the [CustomPeriodGatherer type](https://github.com/openshift/insights-operator/blob/master/pkg/gatherers/interface.go#L21)).
```
Comment (xJustin) [inline docs/arch.md:288]: @tremes I just noticed your edit duplicates "three types of gatherers (see below)" -was this intentional?

Comment (tremes) [reply docs/arch.md:288]: Ohhh nice catch! Thanks.

File: docs/arch.md [hunk 8]
```diff
@@ -285,30 +285,28 @@ Such a client is used in [GatherMachineSet](pkg/gather/clusterconfig/clusterconf
...
-- `important` meaning if that gather-function has an error we will notify `periodic.go` about it, which will handle it accordingly.
-- `failable` meaning if that gather-function has an error we will just log it and add it to our metadata.
-This is necessary as we are expanding into gathering data about resources that are not guaranteed to be present on the cluster. By default if a resource is not present we shouldn't see an error, but it's better to be safe.
+Insights operator defines three types of gatherers (see below). Each of them must implement the [Interface](https://github.com/openshift/insights-operator/blob/master/pkg/gatherers/interface.go#L11) and they are initialized by calling [gather.CreateAllGatherers](https://github.com/openshift/insights-operator/blob/master/pkg/controller/operator.go#L124) in `operator.go`. The actual gathering is triggered in [periodic.go](https://github.com/openshift/insights-operator/blob/master/pkg/controller/periodic/periodic.go#L93), but not every gatherer is triggered every time ( for example, see the [CustomPeriodGatherer type](https://github.com/openshift/insights-operator/blob/master/pkg/gatherers/interface.go#L21)).
```
Comment (natiiix) [inline docs/arch.md:288]: The only thing I could find is the extra space after an opening parenthesis.

**Review by natiiix — APPROVED** [review]
I see no significant issues with the new text.
/lgtm

**Comment by tremes** [comment]
Applying remaining labels myself since it's minor documentation change.
/label qe-approved
/label docs-approved

**Pull request merged**